agent/installer/installer.go

```diff
@@ -33,27 +33,38 @@ type installer struct {
 
 // getSupportedRegistry returns a registry with installers for the supported OS and K8s
 func getSupportedRegistry(bd *bundleDownloader, ob algo.OutputBuilder) registry {
-	var supportedOsK8s = []struct {
-		os   string
-		k8s  string
-		algo algo.K8sStepProvider
-	}{
-		{"Ubuntu_20.04.1_x86-64", "v1.22.1", &algo.Ubuntu20_4K8s1_22{}},
+	reg := newRegistry()
+
+	addBundleInstaller := func(osBundle, k8sBundle string, stepProvider algo.K8sStepProvider) {
+		a := algo.BaseK8sInstaller{
+			K8sStepProvider: stepProvider,
+			/* BundlePath: will be set when tag is known */
+			OutputBuilder: ob}
+
+		reg.AddBundleInstaller(osBundle, k8sBundle, &a)
+	}
+
+	{
+		// Ubuntu
+
+		// BYOH Bundle Repository. Associate bundle with installer
+		linuxDistro := "Ubuntu_20.04.1_x86-64"
+		addBundleInstaller(linuxDistro, "v1.22.1", &algo.Ubuntu20_4K8s1_22{})
 		/*
-		 * ADD HERE to add support for new os or k8s
-		 * You may map new versions to old classes if they do the job
+		 * PLACEHOLDER - ADD MORE K8S VERSIONS HERE
 		 */
-	}
 
-	reg := NewRegistry()
-	for _, t := range supportedOsK8s {
-		a := &algo.BaseK8sInstaller{
-			K8sStepProvider: t.algo,
-			/*BundlePath: will be set when tag is known */
-			OutputBuilder: ob}
-		reg.Add(t.os, t.k8s, a)
+		// Match concrete os version to repository os version
+		reg.AddOsFilter("Ubuntu_20.04.*_x86-64", linuxDistro)
```

Comment:
Question - do we need to hard code this? Can this be injected/overridden via a config file? This could help in our integration testing too.

Reply:
It can be made configurable when needed.

Technically speaking
- only the "Ubuntu_20.04.*_x86-64" can be configurable
- the "Ubuntu_20.04.1_x86-64" cannot be configurable because it is bound to a struct which may be different for OSes and versions. It is the core of the installer

Workflow speaking,
- [yixingjia](https://github.com/vmware-tanzu/cluster-api-provider-bringyourownhost/pull/207#discussion_r743401846) implies that this lib is the source of truth and needs to be queried
- if it is configurable whose responsibility would be to do that?

Testing speaking,
- my experience so far is that we need to override the detection of the current OS. E.g. if lib has ubuntu2004 and I'm running on ubuntu1804, I can override and treat it as ubuntu2004

Reply:
Had a meeting with Dharmjit and Yixing where config files were also mentioned but regarding that downstream version for k8s is 1.2.1+vmware.1.
Agreed we need to look at this holistically post FC.

Reply:
Okay for now, but we need to revisit this again.